File: swift-bridge/Sources/SceneKitBridge/Scene.swift
import Foundation
import SceneKit

@_cdecl("scn_scene_new")
public func scn_scene_new() -> UnsafeMutableRawPointer? {
    scnRetain(SCNScene())
}

@_cdecl("scn_scene_new_named")
public func scn_scene_new_named(_ name: UnsafePointer<CChar>?) -> UnsafeMutableRawPointer? {
    guard let name else { return nil }
    guard let scene = SCNScene(named: String(cString: name)) else { return nil }
    return scnRetain(scene)
}

@_cdecl("scn_scene_new_url")
public func scn_scene_new_url(
    _ path: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let path else { return nil }
    do {
        let scene = try SCNScene(url: URL(fileURLWithPath: String(cString: path)), options: nil)
        return scnRetain(scene)
    } catch {
        outError?.pointee = scnDup(error.localizedDescription)
        return nil
    }
}

@_cdecl("scn_scene_root_node")
public func scn_scene_root_node(_ sceneHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let scene: SCNScene = scnBorrow(sceneHandle) else { return nil }
    return scnRetain(scene.rootNode)
}

@_cdecl("scn_scene_background")
public func scn_scene_background(_ sceneHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let scene: SCNScene = scnBorrow(sceneHandle) else { return nil }
    return scnRetain(scene.background)
}

@_cdecl("scn_scene_lighting_environment")
public func scn_scene_lighting_environment(_ sceneHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let scene: SCNScene = scnBorrow(sceneHandle) else { return nil }
    return scnRetain(scene.lightingEnvironment)
}

@_cdecl("scn_scene_set_fog_color")
public func scn_scene_set_fog_color(
    _ sceneHandle: UnsafeMutableRawPointer?,
    _ r: Float,
    _ g: Float,
    _ b: Float,
    _ a: Float
) {
    guard let scene: SCNScene = scnBorrow(sceneHandle) else { return }
    scene.fogColor = scnMakeColor(r: r, g: g, b: b, a: a)
}

@_cdecl("scn_scene_copy_fog_color")
public func scn_scene_copy_fog_color(
    _ sceneHandle: UnsafeMutableRawPointer?,
    _ outRGBA: UnsafeMutableRawPointer?
) -> Bool {
    guard let scene: SCNScene = scnBorrow(sceneHandle) else { return false }
    return scnWriteColor(scene.fogColor, outRGBA: outRGBA)
}

File: src/scene.rs
use std::path::Path;

use crate::color::Color;
use crate::error::{take_error, SceneKitError};
use crate::ffi;
use crate::material::MaterialProperty;
use crate::node::Node;
use crate::private::{cstring_from_path, cstring_from_str, handle_type};

handle_type!(Scene);

impl Scene {
    /// Creates a wrapped `SCNScene` instance.
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_scene_new()) }
    }

    /// Mirrors `SCNScene.named`.
    #[must_use]
    pub fn named(name: &str) -> Option<Self> {
        let name = cstring_from_str(name)?;
        unsafe { Self::from_raw(ffi::scn_scene_new_named(name.as_ptr())) }
    }

    /// Mirrors `SCNScene.fromUrl`.
    pub fn from_url(path: impl AsRef<Path>) -> Result<Self, SceneKitError> {
        let path = cstring_from_path(path.as_ref())
            .ok_or_else(|| SceneKitError::new("path contains an interior NUL byte"))?;
        let mut error = core::ptr::null_mut();
        let ptr = unsafe { ffi::scn_scene_new_url(path.as_ptr(), &mut error) };
        if ptr.is_null() {
            Err(unsafe { take_error(error, "SCNScene(url:) returned nil") })
        } else {
            Ok(unsafe { Self::from_raw_unchecked(ptr) })
        }
    }

    /// Mirrors `SCNScene.rootNode`.
    #[must_use]
    pub fn root_node(&self) -> Node {
        unsafe { Node::from_raw_unchecked(ffi::scn_scene_root_node(self.ptr)) }
    }

    /// Mirrors `SCNScene.background`.
    #[must_use]
    pub fn background(&self) -> MaterialProperty {
        unsafe { MaterialProperty::from_raw_unchecked(ffi::scn_scene_background(self.ptr)) }
    }

    /// Mirrors `SCNScene.lightingEnvironment`.
    #[must_use]
    pub fn lighting_environment(&self) -> MaterialProperty {
        unsafe {
            MaterialProperty::from_raw_unchecked(ffi::scn_scene_lighting_environment(self.ptr))
        }
    }

    /// Sets the `SCNScene.fogColor` member.
    pub fn set_fog_color(&self, color: Color) {
        unsafe { ffi::scn_scene_set_fog_color(self.ptr, color.r, color.g, color.b, color.a) };
    }

    /// Mirrors `SCNScene.fogColor`.
    #[must_use]
    pub fn fog_color(&self) -> Option<Color> {
        let mut rgba = [0.0_f32; 4];
        let ok = unsafe { ffi::scn_scene_copy_fog_color(self.ptr, rgba.as_mut_ptr().cast()) };
        ok.then(|| Color::rgba(rgba[0], rgba[1], rgba[2], rgba[3]))
    }
}
